core/runner/src/main/java/com/cognifide/aet/runner/processing/data/SuiteComparator.java

```diff
@@ -0,0 +1,226 @@
+
+    boolean firstHasUniqueTests =
+        Sets.difference(new HashSet<>(firstTestsNames),
+            new HashSet<>(secondTestsNames)).size() != 0;
+    boolean secondHasUniqueTests = Sets.difference(new HashSet<>(secondTestsNames),
+        new HashSet<>(firstTestsNames)).size() != 0;
+
+    boolean suitesHaveDifferentTests = firstHasUniqueTests || secondHasUniqueTests;
+
+    return suitesHaveDifferentTestsCount || suitesHaveDifferentTests;
+  }
+
+  private static void compareSteps(Suite first, Suite second, List<String> comparisonMessages) {
+    if (isNullOrEmpty(first.getTests()) || isNullOrEmpty(first.getTests())) {
+      return;
+    }
+
+    first.getTests().forEach(firstTest -> {
+      Optional<Test> secondTest = second.getTests().stream()
+          .filter(found -> found.getName().equals(firstTest.getName()))
+          .findFirst();
+
+      if (secondTest.isPresent()) {
+        if (testsHaveDifferentSteps(firstTest, secondTest.get())) {
+          comparisonMessages.add(String.format(
+              "Tests: %s in suite: %s, with id: %s and suite: %s, with id: %s have different structure.",
+              firstTest.getName(), first.getName(), first.getCorrelationId(),
+              second.getName(), second.getCorrelationId()));
+        }
+      }
+    });
+  }
+
+  private static boolean testsHaveDifferentSteps(Test first, Test second) {
+    if (isNullOrEmpty(first.getUrls()) && !isNullOrEmpty(second.getUrls())) {
+      return true;
+    }
+
+    Collection<List<Url>> urlPairs = Stream
+        .concat(first.getUrls().stream(), second.getUrls().stream())
+        .collect(Collectors.groupingBy(Url::getName))
+        .values();
+
+    if (anyHasMissingPair(urlPairs)) {
+      return true;
+    }
+
+    List<Url> firstUrlPair = urlPairs.iterator().next();
+
+    return stepsHaveDifferentCollectorsOrComparators(firstUrlPair.get(0).getSteps(),
+        firstUrlPair.get(1).getSteps());
+  }
+
+  private static boolean stepsHaveDifferentCollectorsOrComparators(
+      List<Step> firstStep, List<Step> secondStep) {
+    return areCollectionsDifferent(firstStep, secondStep, SuiteComparator::areOperationsDifferent);
+  }
+
+  private static <T> boolean areCollectionsDifferent(Collection<T> first, Collection<T> second,
+      BiFunction<Collection<T>, Collection<T>, Boolean> compareDifference) {
+    if (isNullOrEmpty(first, second)) {
+      return false;
+    }
+
+    if (isNullOrEmpty(first) ^ isNullOrEmpty(second) || first.size() != second.size()) {
+      return true;
+    }
+
+    return compareDifference.apply(first, second);
+  }
+
+  private static <T extends Operation> boolean areOperationsDifferent(Collection<T> first,
+      Collection<T> second) {
+    return first.size() != second.size() ||
+        forEachFindFirst(first, second,
+            (operation1, operation2) -> {
+              boolean typeAndParamsDifferent = !areTypeAndParamsEqual(operation1, operation2);
```

Comment:
This could be extracted to separate method. It's hard to understand what's going on here.